Allow users to pass in a custom color for background fill

itbra opened the pull request:
##### This patch currently works, but is work in progress.

Please follow the discussion below!
#### Summary of Changes

Currently the color to use for non-transparent images' background fill is hardcoded to `black` which provides no flexibility. Allowing the users to define a desired background color fixes this.
This PR adds two new functions `setFillColor` and `getFillColor` as well as two new helper functions and changes the functions `resize` and  `rotate` to use the desired fillColor, which defaults to `black` preserving BC.
#### Testing Instructions
- [ ] Create a footage array of test files for each image file type.
  I recomment downloading my test files from the links shown below into a folder named 'testfiled' in your Joomla installation's tmp directory. These are specially prepared for this test scenario.
- [ ] Copy and paste the code below to whichever file you'd like to use for running the test. I used my component's entry file.

Example:

```
$files  = array(
   'GIF' => array(
      'NO transparency in input file'   => JPATH_ROOT . '/tmp/testfiles/koZs44Z.gif',  // http://i.imgur.com/koZs44Z.gif
      'WITH transparency in input file' => JPATH_ROOT . '/tmp/testfiles/kyxRLmz.gif'  // http://i.imgur.com/kyxRLmz.gif
   ),
   'PNG' => array(
      'NO transparency in input file'   => JPATH_ROOT . '/tmp/testfiles/FcoS851.png',  // http://i.imgur.com/FcoS851.png
      'WITH transparency in input file' => JPATH_ROOT . '/tmp/testfiles/VL9BgDi.png'  // http://i.imgur.com/VL9BgDi.png
   ),
   'JPG' => array(
      'NO transparency in input file'   => JPATH_ROOT . '/tmp/testfiles/TFnVnnF.jpg'   // http://i.imgur.com/TFnVnnF.jpg
   )
);

// Font styles for the renderer
$ftColors = array(
   'none'        => 'rgb(0,0,0)',
   'red'         => 'rgb(255,0,0)',
   'green'       => 'rgb(0,255,0)',
   'blue'        => 'rgb(0,0,255)',
   'azure'       => 'rgb(0,255,255)',
   'pink'        => 'rgb(255,0,255)',
   'yellow'      => 'rgb(255,255,0)',
   'white'       => 'rgb(255,255,255)',
   'transparent' => 'rgb(0,0,0)',
   // 'nonsense' => 'rgb(0,0,0)'
);

// Fill colors to pass to JImage
$bgColors = array(
   'none'        => array(),
   'red'         => array('red'   =>    255,                                   'alpha' =>  33),
   'green'       => array(                   'green' => 255,                   'alpha' =>  66),
   'blue'        => array(                                    'blue' =>  255,  'alpha' =>  99),
   'azure'       => array(                   'green' => 255,  'blue' =>  255,  'alpha' =>  33),
   'pink'        => array('red'   =>    255,                  'blue' =>  255,  'alpha' =>  66),
   'yellow'      => array('red'   =>    255, 'green' => 255,                   'alpha' =>  99),
   'white'       => array('red'   =>    255, 'green' => 255,  'blue' =>  255),
   'transparent' => array(                                                     'alpha' =>  99),
   // 'nonsense' => array('red'   => 'this', 'green' => 'is', 'blue' => 'all', 'alpha' => 'nonsense, but it works')
);

// Render image
$render = function($path, $extension, $title)
{
   // Define neutral background color for every image to show the different fill colors don't apply when there's transparency
   $neutral = 'rgb(128,128,128)';

   echo '<div style="height:310px; background:' . $neutral . '; border:1px solid black; float:left;">';
   echo $title;
   echo '<img src="' . $path . '" alt="New Image" style="border-top:1px solid black;" />';
   echo '</div>';
};

array_walk($files, function(&$inFiles, $type) use(&$render, &$bgColors, &$ftColors)
{
   array_walk($inFiles, function(&$inFile, &$transparencyType) use(&$type, &$render, &$bgColors, &$ftColors)
   {
      echo '<h2 style="clear:both; padding-top:1em">Testing image type ' . $type . ' <small>(' . $transparencyType . ')</small></h2>';

      $pathinfo = pathinfo($inFile);

      $scaleFmt = JImage::SCALE_FIT;
      $outPath  = JPATH_ROOT . '/tmp/';
      $outFile  = null;
      $outExt   = $pathinfo['extension'];
      $outType  = explode('/', getimagesize($inFile)['mime']);
      $outType  = strtoupper( array_pop($outType) );
      $outType  = $outType === 'GIF' ? IMAGETYPE_GIF : ($outType === 'PNG' ? IMAGETYPE_PNG : IMAGETYPE_JPEG);
      $outFmt   = array('width' => '420', 'height' => '240');
      $outOpts  = array('quality' => 100);

      $i = 1;
      foreach ($bgColors as $color => &$rgb)
      {
         // Resize image defining no fill color (should fall back to BLACK):
         $outFile = "{$outPath}{$pathinfo['filename']}-{$outFmt['width']}x{$outFmt['height']}-{$color}.{$outExt}";
         $jimage  = new JImage($inFile);
         $jimage
         ->setFillColor($rgb)
         ->resize($outFmt['width'], $outFmt['height'], true, $scaleFmt)
         ->toFile($outFile, $outType, $outOpts);

         $render(
            str_replace(JPATH_ROOT, '', $outFile),
            $outExt,
           '<pre style="width:100%; text-align:center; color:' . $ftColors[$color] . '"><strong>' .
            print_r("TEST {$i} ==> " . strtoupper($color) . (in_array($color, array('none','nonsense')) ? ' (falls back to BLACK)' : ''), true) .
            '</strong></pre>' .
            '<pre style="width:100%; text-align:center; color:' . $ftColors[$color] . '"><strong>fillColor: ' .
            preg_replace('/(\}|\{)/', '', str_replace(',', ', ', str_replace('"', '', json_encode($jimage->getFillColor())))) .
            '<strong></pre>'
         );

         $i += 1;
      }
   });
});

echo '<pre style="clear:both; padding-top:1em">' . print_r('Imagetest done!', true) . '</pre>';
jexit();
```

This will render a collection of files for every image from the footage collection.

Please report issues and code improvement suggestions!

itbra committed:
Allow users to pass in a custom color for background fill

Currently the color to use for non-transparent images' background fill is hardcoded to `black` which provides no flexibility. Allowing the users to define a desired background color fixes this.
This PR adds two new functions `setFillColor` and `getFillColor` and changes the function `resize` to use the defined fillColor, which defaults to `black`.

itbra committed:
Fix code style issues

Attempt to fix code style issues that cause one of the tests to fail. These issues were:
```
686 | ERROR | Expected 2 spaces after the longest type
686 | ERROR | Expected 2 spaces after the longest variable name
701 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line
714 | ERROR | Closing parenthesis of a multi-line function call must be on a line by itself
721 | ERROR | Increment operators should be used where possible; found "$i += 1)" but expected "$i++"
```

itbra committed:
Fix code style issue

Fix code style issues
Attempt to fix code style issue that caused one of the tests to fail. These issue was:
```
701 | ERROR | Opening parenthesis of a multi-line function call must be the last content on the line
```

itbra committed:
Fix code style issue

Fix code style issue
Fix code style issues
Attempt to fix code style issue that caused one of the tests to fail. These issue was:
```
702 | ERROR | Whitespace found at end of line
713 | ERROR | Whitespace found at end of line
```

bembelimen commented on libraries/joomla/image/image.php:
```diff
@@ -727,7 +795,10 @@ public function resize($width, $height, $createNew = true, $scaleMethod = self::
 			{
-				$transparency = imagecolorAllocateAlpha($this->handle, 0, 0, 0, 127);
+				// Get fill color.
+				$fillColor = $this->getFillColor();
```
How do you make sure, that $fillColor has values (e.g. if setFillColors() was not called/set before)?

itbra replied:
You are right. I forgot to take care of that yet and will fix it soon. I am still working on this topic and this file since I encountered more places in `JImage` where the default color has been hardcoded and I noticed that I have to handle all places where `$this->isTransparent()` returns true even for none-transparent files such as `JPEG`s.

Meanwhile all suggestions are highly welcome!

bembelimen replied:
Perhaps it make sense to use specific keys like ['red'], ['blue'], ['green'] and handle ```$this->fillColor``` as object (->red, ->blue,->green) + fill in some default values in the constructor

itbra replied:
In fact I had the same idea, but ended up to not modify the class constructor because of the BC. I don't know how excessively this class is used throughout the CMS. For that reason I decided to introduce these two new functions with the setter allowing to chain other functions. Calling `setFillColor` appeared less intrusive to me and can easily be added by those who require to have this flexibility and who know where to add it.

According the useage of an object over an array I had the same idea, but ended up passing an indexed array for the sake of readability and not blowing the code to write to set the fill color. It is far less code and better readable when I write
```
$jimage
->setFillColor(array(255))->toFile( ... );
```
than if I had to write
```
$jimage
->setFillColor((object) array('red' => 255, 'green' => 128, 'blue' => 64, 'alpha' => 96))
->toFile( ... );
```
However, using the assoc array with the color(s) exactly mentioned would be clearer for the further preparation and usage by `JImage`. I think I should try both ways when I'll put my hands on it again, which won't happen this week since I'm out of time to spend on it.

RonakParmar commented:
@itbra  Hi, please provide some more testing information.

I have applied this patch in latest staging site in my local and got warnings, please see the attached screen-dump.
I have written below code in my view default.php file, let me know If I am doing anything wrong.
$jimage = new JImage('/path/to/image/file.jpg'); (I have replaced this with original image path)

// Resize image using default fill color (black):
$jimage
->resize(320, 240, true, JImage::SCALE_FIT);

// Resize image using red as fill color:
$jimage
->setFillColor(array(255))   // same as 255,0,0 - missing values are always filled from default value
->resize(320, 240, true, JImage::SCALE_FIT);
die;<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10268">issues.joomla.org/joomla-cms/10268</a>.</sub>

itbra commented:
@RonakParmar please be patient, I am currently working in this class and there's actually more to handle than initially detected. As soon as my copy works smoothly I'll update the PR and the testing instructions (as much as necessary).

itbra committed:
Code update

Previous code has been updated in favour of discussion with @bembelimen.

itbra committed:
Code cleanup

Forgot to remove debug comments.

itbra committed:
code cleanup

Forgot to remove debug information

itbra committed:
Fix code style issues

itbra committed:
Fix code style issues

itbra committed:
Minor changes

Changed modifier of functions `fixFillColor` and `getColorForIndex` since they must not be public.
Removed quality factor variable in `toFile` when input file is of type GIF, because its not rquired for gifs.

itbra committed:
Variable type fix + new getter

Fixed type of new variable $type from `string` to `int` and introduced new getter for this property allowing to get the loaded image's type immediately.

itbra committed:
Revert camel case notation of image processing functions

itbra committed:
Change default transparency value in setFillColor()

Transparency must not be 127 initially.

itbra commented:
Finally this patch works (at least this is what my test setup reflects). Yet there's one issue which I'm wondering whether this is an issue of understanding or implementation. PNGs and GIFs are always rendered with areas that are not part of the image content getting filled with transparency. But, must it not be assumed that event these types of images may have a their non-image areas filled with the defined fillColor like is done with JPEGs? If this is the case then I need some assistance as I can't seem to find the code responsible to use transparency no matter what fillColor is defined.

Or am I on the wrong way and transparency goes always over a defined fillColor for PNGs and GIFs?

itbra committed:
Code cleanup

itbra committed:
Rename fixFillColor() to sanitizeFillColor()

itbra committed:
Add change to be discussed regarding fillColor over transparency

bembelimen commented:
Hi @itbra could you provide an example image/code, I'm not sure, if I understand you correctly

itbra commented:
Hold on, I'm gonny create a "allover screenshot" of the output currently rendered with the defined test scenario.

itbra committed:
Update image.php

itbra committed:
Fix code style issue

CiverBlack commented:
Hello, you are saying you are providing a link with test images , but i can not find that link. Could you please provide it. I would also agree that the transparency should be filled with the defined fillColor.

@iCampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10268">issues.joomla.org/tracker/joomla-cms/10268</a>.</sub>

franz-wohlkoenig commented:
If this Issue get no Response, it will be closed at 28th February 2018.

franz-wohlkoenig commented:
This has been closed due to lack of response to the requests above – it can always be reopened.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10268">issues.joomla.org/tracker/joomla-cms/10268</a>.</sub>
